### DISCUSSION TITLE
Getting Current User Email and Current User OneDrive ID to use in Handlebars Expression

### DISCUSSION THREAD

**Discussion by dontthroworanges** [microsoft-search/pnp-modern-search#3779] [Q&A]
I am working on a way to trigger the "Open in App" functionality from within the Search Results web part. This would be very helpful since we work in a lot of PDFs, and it would be nice to be able to bypass the web-based PDF previewer if the user wants to jump straight into Adobe Acrobat on their desktop. 

Grabbing the "odopen:" URL and figuring out the various file properties to use was pretty straight forward. However, the URL also requires the current user's email and what I believe is a OneDrive ID. Both of these strings need to correspond with the users OneDrive desktop application to allow the handoff of the file. Otherwise you'll get an error or OneDrive on the desktop will start the Add Account Wizard. 

Is it possible to get these two properties somehow with the webparts? Or maybe another means?

Here is a sample URL:

odopen://openFile/?fileId=7495fb9f-b000-7a3c-a3d9-f5ef01f55297&siteId={e355922a-2b81-4d4a-a6987-a68587aa7089}&listId={45e85ecb-3170-496c-bb86-c0ccccd2d7cb}&userEmail=email@contoso.com&userId=1a201545-13f4-4879-859a-90b27b164ae7&webUrl=https://contoso.sharepoint.com/sites/SiteName&fileName=Document Name.pdf

**Comment by dontthroworanges** [comment]
I think that Tokens might be the answer... However, dropping these values into the URL leaves me with a dead icon. Can anyone spot the issue?

    <a href="odopen://openFile/?fileId={{slot item @root.slots.ListItemID}}&siteId={{slot item @root.slots.SiteID}}&listId={{slot item @root.slots.ListID}}&userEmail={User.email}&userID={User.\userprofile_guid}=&webUrl={{slot item @root.slots.SPWebUrl}}&fileName={{slot item @root.slots.Filename}}" target="_blank" data-interception="off" style="color: {{@root.theme.semanticColors.link}}; text-decoration: none; margin-right: 12px;" title="Open in Desktop Application">
        <pnp-icon data-name="OpenInNewTab" aria-hidden="true" style="font-size: 16px;"></pnp-icon>
    </a>

**Reply by patrikhellgren** [reply]
The {User.\<property>} tokens are not parsed when used in Handlebars templates, see [Where can I use tokens?](https://microsoft-search.github.io/pnp-modern-search/usage/search-results/tokens/#where-can-i-use-tokens) for more information.


As a sidenote, as you might already know, if you wanted to accomplish this for Office files you could use [Office URI Schemes](https://learn.microsoft.com/en-us/office/client-developer/office-uri-schemes). Using that the URL would be as simple as ms-word:ofe|u|https<span>://</span>contoso.sharepoint.com/sites/PnPModernSearch/Shared Documents/Document1.docx, i.e. the Handlebars syntax for that would be ms-word:ofe|u|{{item.Path}}. You would of course need to handle the different scheme-names for the different file types but that should be quite easy using regular Handlebars syntax.
There doesn't seem to be a URI scheme for Adobe Acrobat that would work with editing files in SharePoint.

**Reply by patrikhellgren** [reply]
You could use the following Handlebars expressions to get the properties you wanted:

{{getPageContext "user.email"}}
{{getPageContext "aadInfo.userId"}}

For information on ```getPageContext``` see [https://microsoft-search.github.io/pnp-modern-search/extensibility/templating/#getpagecontext](https://microsoft-search.github.io/pnp-modern-search/extensibility/templating/#getpagecontext)

To view the full page context you can execute the following in the browsers developer console:
```window.spModuleLoader._bundledComponents["b6917cb1-93a0-4b97-a84d-7cf49975d4ec"].PageManager._instance.pageContext```

**Reply by dontthroworanges** [reply]
@patrikhellgren Thanks for all the info! I'm definitely a novice when it comes to all of this, so I appreciate all the help. 

Unfortunately, I'm still hitting a brick wall and can't get the link to be recognized. Whatever I seem to do, using the odopen protocol in href leads to an unclickable link. Anyone else want to try this and see if it's just me or something else?

`<div style="display: flex; align-items: center;">
    <a href="odopen://openFile/?fileId={{@root.NormListID}}&siteId={{@root.NormSiteID}}&listId={{@root.NormListID}}&userEmail={{getPageContext "user.email"}}&userID={{getPageContext "aadInfo.userId"}}&webUrl={{@root.SPWebUrl}}&fileName={{@root.Filename}}" target="_blank" data-interception="off" style="color: {{@root.theme.semanticColors.link}}; text-decoration: none; margin-right: 12px;" title="Open in Desktop Application">
        <pnp-icon data-name="OpenInNewTab" aria-hidden="true" style="font-size: 16px;"></pnp-icon>
    </a>
</div>`

**Reply by patrikhellgren** [reply]
@dontthroworanges Your issue is not with the properties but with the odopen protocol handler which is being stripped from the parsed html. See #3415 for more info. It would be easy to make the change to handle odopen also but I can not get it to open the file in the app so it would be of no use if it does not work.
You also have a few errors in your syntax, all `@root.` except `@root.theme...` should be changed to `item.` instead. You can test the opening in app by just specifing https instead of odopen in your Handlebar expression and then use your browsers dev tools to edit the href to change the https to odopen for one of the links and then click the link to see if it opens in the app.
